Replace the use of `alldocs` with respective collections for record retrieval in NCBI exporter

sujaypatil96 opened the pull request:
Fixes https://github.com/microbiomedata/issues/issues/1063

TODO:

- [x] Make sure the export code picks up on _platform_ and _instrument_model_ Attributes correctly
- [x] For data object records that are linked by a mainfest record, the SRA Action block needs to have references to all the connected files

<!--┌─────────────────────────────────────────────────────────────────────────┐
    │ 1. Summary (required)                                                   │
    │                                                                         │
    │ Summarize the changes you made on this branch. This is typically a more │
    │ detailed restatement of the PR title.                                   │
    │                                                                         │
    │ Example: "On this branch, I updated the `/studies/{study_id}` endpoint  │
    │           so it returns an HTTP 404 response when the specified study   │
    │           does not exist."                                              │
    └─────────────────────────────────────────────────────────────────────────┘-->

On this branch, I...

### Details

<!--┌─────────────────────────────────────────────────────────────────────────┐
    │ 2. Details (optional)                                                   │
    │                                                                         │
    │ Provide additional information you think readers will find useful.      │
    │ Readers include PR reviewers, release note authors, app debuggers, and  │
    │ your future self. Additional information might include motivation,      │
    │ rationale, and a description of how things used to be.                  │
    │                                                                         │
    │ Example: "It previously returned an HTTP 404 response and an empty      │
    │           JSON object."                                                 │
    └─────────────────────────────────────────────────────────────────────────┘-->

...

### Related issue(s)

<!--┌─────────────────────────────────────────────────────────────────────────┐
    │ 3. Related issue(s) (optional)                                          │
    │                                                                         │
    │ Link to any GitHub issue(s) this branch was designed to resolve.        │
    │                                                                         │
    │ Example: "Fixes #12345"                                                 │
    └─────────────────────────────────────────────────────────────────────────┘-->

...

### Related subsystem(s)

<!--┌─────────────────────────────────────────────────────────────────────────┐
    │ 4. Related subsystem(s) (required)                                      │
    │                                                                         │
    │ Mark the checkbox next to each subsystem related to the changes in this │
    │ branch. This information might influence who you request reviews from.  │
    │                                                                         │
    │ Example: If you modified the `/studies/{study_id}` API endpoint,        │
    │          mark the checkbox next to "Runtime API (except the Minter)".   │
    └─────────────────────────────────────────────────────────────────────────┘-->

- [ ] Runtime API (except the Minter)
- [ ] Minter
- [ ] Dagster
- [ ] Project documentation (in the `docs` directory)
- [x] Translators (metadata ingest pipelines)
- [ ] MongoDB migrations
- [ ] Other

### Testing

<!--┌─────────────────────────────────────────────────────────────────────────┐
    │ 5. Testing (required)                                                   │
    │                                                                         │
    │ Indicate whether you have already tested the changes this branch        │
    │ contains; and, if so, how someone other than you can test them. That    │
    │ may involve attaching example files or ad hoc test instructions.        │
    │                                                                         │
    │ Example: "I tested these changes by adding a pytest test that ensures   │
    │           the database does not contain a Study whose ID is `foo`,      │
    │           then submits an HTTP request to `/studies/foo` and confirms   │
    │           the response status is 404."                                  │
    └─────────────────────────────────────────────────────────────────────────┘-->

- [x] I tested these changes (explain below)
- [ ] I did not test these changes

I tested these changes by...

### Documentation

<!--┌─────────────────────────────────────────────────────────────────────────┐
    │ 6. Documentation (required)                                             │
    │                                                                         │
    │ Indicate whether, in this branch, you have updated all documentation    │
    │ that would otherwise become inaccurate if this branch were to be        │
    │ merged in.                                                              │
    └─────────────────────────────────────────────────────────────────────────┘-->

- [ ] I **have not checked** for relevant documentation yet (e.g. in the `docs` directory)
- [ ] I have **updated** all relevant documentation so it will remain accurate
- [ ] Other (explain below)

### Maintainability

<!--┌─────────────────────────────────────────────────────────────────────────┐
    │ 7. Maintainability (required)                                           │
    │                                                                         │
    │ Indicate whether you have done each of these things that can make code  │
    │ easier to maintain, whether by your teammates or by your future self.   │
    └─────────────────────────────────────────────────────────────────────────┘-->

- [ ] Every Python function I defined includes a docstring _(test functions are exempt from this)_
- [x] Every Python function parameter I introduced includes a type hint (e.g. `study_id: str`)
- [ ] All "to do" or "fix me" Python comments I added begin with either `# TODO` or `# FIXME`
- [x] I used `black` to format all the Python files I created/modified
- [x] The PR title is in the imperative mood (e.g. "Do X") and not the declarative mood (e.g. "Does X" or "Did X")

eecavanna commented:
I'll review this PR this morning.

eecavanna commented on nmdc_runtime/site/export/ncbi_xml_utils.py:
```diff
@@ -45,35 +47,41 @@ def get_instruments(instrument_set_collection):
 
-def fetch_data_objects_from_biosamples(all_docs_collection, biosamples_list):
+def fetch_data_objects_from_biosamples(
+    all_docs_collection, data_object_set, biosamples_list
```
I recommend adding type hints to function parameters and return values, and docstrings to functions. I recommend these things be done to all functions added in the diff.

eecavanna commented on nmdc_runtime/site/export/ncbi_xml_utils.py:
```diff
@@ -105,11 +115,13 @@ def fetch_nucleotide_sequencing_from_biosamples(all_docs_collection, biosamples_
                         )
                         if nucleotide_sequencing_doc:
-                            collected_data_objects.append(nucleotide_sequencing_doc)
+                            collected_data_objects.append(
+                                strip_oid(nucleotide_sequencing_doc)
+                            )
```
I'm confused about the variable names here. I was surprised to see a "nucleotide sequencing" thing being stored in a list of "data objects". I think neither of those things inherits from ("is a") the other, according to the schema.

eecavanna commented on nmdc_runtime/site/export/ncbi_xml.py:
```diff
@@ -366,7 +366,12 @@ def set_fastq(
                             # Currently, we are making the assumption that only one instrument
                             # is used to sequence a Biosample
-                            instrument_id = ntseq.get("instrument_used", "")[0]
+                            instrument_used = ntseq.get("instrument_used", "")
+                            if not instrument_used:
+                                instrument_id = None
+                            else:
+                                instrument_id = instrument_used[0]
```
I think `instrument_used` can be either a list or a string. This is one situation where I'd recommend adding a type hint to the code (on line 369). Either `Union[str, list]` or `str | list`, depending on what Python version is being used. 

Alternatively, instead of using an empty string to represent the "null" value, one could use `[]`. That way, the value would always be a list, not a "string or list".

eecavanna reviewed (approved):
Thanks for making these changes to accommodate the new sparseness of the "all docs" (maybe one day renamed to "all refs" or something) collection, @sujaypatil96.

I made some comments related to type hints, docstrings, and 1-2 variable names.

sujaypatil96 commented:
Thank you for the review comments @eecavanna, they were very helpful! I was being lazy by not adding type hints and docstrings. I've done that now. I think I've addressed all your review comments, feel free to drop another review.

eecavanna commented on nmdc_runtime/site/export/ncbi_xml.py:
```diff
@@ -366,7 +366,14 @@ def set_fastq(
                             # Currently, we are making the assumption that only one instrument
                             # is used to sequence a Biosample
-                            instrument_id = ntseq.get("instrument_used", "")[0]
+                            instrument_used: Union[str, List[str]] = ntseq.get(
+                                "instrument_used", []
+                            )
+                            if not instrument_used:
+                                instrument_id = None
+                            else:
+                                instrument_id = instrument_used[0]
```
Now that the fallback value is an empty list (instead of an empty string), I think (without knowing where this object comes from) the value will be a list of strings in all situations.

```suggestion
                            instrument_used: List[str] = ntseq.get(
                                "instrument_used", []
                            )
                            if not instrument_used:
                                instrument_id = None
                            else:
                                instrument_id = instrument_used[0]
```

eecavanna commented:
Thanks for making those changes, @sujaypatil96! I found the docstrings and type hints helpful in keeping track of what is happening in the code (while reading it).

I left one follow-up comment about a type hint that I thought was more broad than necessary. I'll defer to you about that. You can merge already or edit-and-merge.
